core/services/ocr2/plugins/directrequestocr/reporting.go

```diff
@@ -0,0 +1,223 @@
+package directrequestocr
+
+import (
+	"context"
+	"encoding/hex"
+
+	uuid "github.com/satori/go.uuid"
+	"github.com/smartcontractkit/chainlink/core/services/directrequestocr"
+	"github.com/smartcontractkit/libocr/commontypes"
+	"github.com/smartcontractkit/libocr/offchainreporting2/types"
+	"google.golang.org/protobuf/proto"
+)
+
+type DirectRequestReportingPluginFactory struct {
+	Logger    commontypes.Logger
+	PluginORM directrequestocr.ORM
+	JobID     uuid.UUID
+}
+
+var _ types.ReportingPluginFactory = (*DirectRequestReportingPluginFactory)(nil)
+
+type directRequestReporting struct {
+	logger      commontypes.Logger
+	pluginORM   directrequestocr.ORM
+	jobID       uuid.UUID
+	reportCodec *reportCodec
+}
+
+var _ types.ReportingPlugin = &directRequestReporting{}
+
+// NewReportingPlugin complies with ReportingPluginFactory
+func (f DirectRequestReportingPluginFactory) NewReportingPlugin(types.ReportingPluginConfig) (types.ReportingPlugin, types.ReportingPluginInfo, error) {
+	info := types.ReportingPluginInfo{
+		Name:          "directRequestReporting",
+		UniqueReports: true, // Enforces (N+F+1)/2 signatures. Must match setting in OCR2Base.sol.
+		// TODO move to config
+		// https://app.shortcut.com/chainlinklabs/story/56615/config-for-reporting-plugin
+		Limits: types.ReportingPluginLimits{
+			MaxQueryLength:       10_000,
+			MaxObservationLength: 10_000,
+			MaxReportLength:      10_000,
+		},
+	}
+	codec, err := NewReportCodec()
+	if err != nil {
+		return nil, info, err
+	}
+	plugin := directRequestReporting{
+		logger:      f.Logger,
+		pluginORM:   f.PluginORM,
+		jobID:       f.JobID,
+		reportCodec: codec,
+	}
+	return &plugin, info, nil
+}
+
+// Query() complies with ReportingPlugin
+func (r *directRequestReporting) Query(ctx context.Context, ts types.ReportTimestamp) (types.Query, error) {
+	r.logger.Debug("directRequestReporting Query phase", commontypes.LogFields{
+		"epoch": ts.Epoch,
+		"round": ts.Round,
+	})
+	// TODO add batch size to config
+	// https://app.shortcut.com/chainlinklabs/story/56615/config-for-reporting-plugin
+	const maxBatchSize = 10
+	results, err := r.pluginORM.FindOldestEntriesByState(directrequestocr.RESULT_READY, maxBatchSize)
```

Comment:
If there are no entries, this will still marshal the Query and run the round?

Reply:
The rounds will run even with a nil query and that's fine.

Reply:
I mean we may not want to proceed with other stages of the round and simply return nil to indicate we don't need this round to continue.

Reply:
Returning nil from Query doesn't stop other stages, Observation will still be invoked.